puppet/zulip/files/hooks/post-deploy.d/sentry.hook

```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/bash
+# Arguments: OLD_COMMIT NEW_COMMIT ...where both are `git describe`
+# output or tag names.  The CWD will be the new deploy directory.
+
+set -e
+set -u
+
+if ! grep -q 'SENTRY_DSN' /etc/zulip/settings.py; then
+    echo "sentry: No DSN configured!  Set SENTRY_DSN in /etc/zulip/settings.py"
+    exit 0
+fi
+
+if ! SENTRY_AUTH_TOKEN=$(crudini --get /etc/zulip/zulip-secrets.conf secrets sentry_release_auth_token); then
+    echo "sentry: No release auth token set!  Set sentry_release_auth_token in /etc/zulip/zulip-secrets.conf"
+    exit 0
+fi
+
+if ! SENTRY_ORG=$(crudini --get /etc/zulip/zulip.conf sentry organization); then
+    echo "sentry: No organization set!  Set sentry.organization in /etc/zulip/zulip.conf"
+    exit 0
+fi
+
+if ! SENTRY_PROJECT=$(crudini --get /etc/zulip/zulip.conf sentry project); then
+    echo "sentry: No project set!  Set setry.project in /etc/zulip/zulip.conf"
+    exit 0
+fi
+
+if ! which sentry-cli >/dev/null; then
+    echo "sentry: No sentry-cli installed!"
+    exit 0
+fi
+
+if ! [ -f ./sentry-release ]; then
+    echo "sentry: No Sentry release was made for this deploy, as it was the first"
+    echo "        deploy with the hooks enabled.  Future deploys will create"
+    echo "        Sentry releases."
+    exit 0
+fi
```

Comment:
Does this have any goofy implications for brand new installs? I'm not fully versed on when and where Puppet runs, but would the first release deployed to a brand new box run into this and thus not have proper Sentry integration? That would be a rather unfortunate surprise to run into for someone (possibly ourselves as we horizontally scale out).

Reply:
The only goofiness is that they don't get a deploy marker.  Since the first deploy to a new box is when it's not getting traffic yet, I don't think this is an issue.  Sentry will create an implicit deploy the first time it sees errors reported with a given revision.  Given that, I think I should adjust the default value (in `zproject/sentry.py` to be `zulip-server@{ZULIP_VERSION}` so that in that case it's still a consistent versioning scheme, but it should still Work.

I should note this in the commit message, but there is the general oddity that if a deploy hook changes during a deploy, then you get the _pre-deploy_ hook with the _old_ version, but the _post-deploy_ from the _new_ version.  I don't see a way to use the pre-deploy hook from the new deploy, though:
 - We want the pre-deploy hook to run before we stop the server
 - We need to stop the server before we apply puppet
 - Applying puppet is what updates the pre-deploy hooks

We could come at it from the other angle, and snapshot `/etc/zulip/post-deploy.d` before puppet is applied, and use that pre-puppet snapshot for the post-deploy hooks later.  That seems like it might be a bit complicated, and while more internally-consistent, I'm not sure it's less confusing than the current behaviour.

This is all kinda a side-effect of puppet'ing changes into `/etc/zulip/` which we declare as "the configuration source for puppet" -- so there'a bit of a tail-wagging-the-dog if puppet is writing there, and it may imply that we're Doing it Wrong to be writing those files from puppet at all.  But it's very much more convenient to do that as a deploy technique.

I guess we could [write the hooks with `replace => no`](https://www.puppetcookbook.com/posts/only-create-file-if-absent.html) as a middle ground but that also seems potentially confusing.

Reply:
It's not ideal for the `pre-deploy` hooks to not run from the new version; it's the kind of that that'll eventually burn us.

I think the main option for that would be to declare that `pre-deploy` hooks are not managed via `puppet`, and get installed by a `cp -a` command. (Or the thing we put into `puppet` for the `zulip_ops` project is a symlink into `/home/zulip/deployments/next/...`).

Reply:
I assume you mean both `pre-deploy` and `post-deploy` would get installed via `cp -a`?  It seems weird that only one would be.  And were you envisioning the `cp -a` step to be done by a human?  Because if you expect `upgrade-zulip` to do it, that puts us in the position of having to parse the puppet classes to know which hook files to install as a pre-step.

Good idea or bad idea: Write out the compiled catalog file (a JSON file describing the compiled end-state of the system after parsing all of the Puppet) as a step in the upgrade process, look at that to know what hooks _will_ be installed, and what their contents will be, and use those for the pre-hooks (...merged with the ones on disk, I guess?)  There's a very minor utility to doing this catalog compilation explicitly, since we call `puppet apply` multiple times in the upgrade process, and it re-does that few-second compilation work each time.  This is a little important because it can shave seconds off of the critical period when the server is stopped and we're applying puppet.

But I'm a little leery of taking what is essentially an intermediate Puppet output step and using it to do a mini puppet-apply on just a subset of the files.  But it does solve the chicken-and-egg problem of puppet and hooks installation, in that we can compile the catalog before stopping the server, and thus know what the hooks will be.

There's a minor scuzziness that compiling the catalog requires that we write the puppet "facts" to disk in `/var/lib/puppet/yaml/facts/$(hostname --fqdn).yaml` which might interact poorly with any other puppet processes on the host.   But we can clean it up afterwards, and anything which races and reads it will just get ... the current set of facts, which is nominally a fine and useful thing to find there.

-----

The other direction to go is just document "copy the pre and post hooks from here, we never puppet them" and the sentry class just installs the CLI for you.

Reply:
Actually, I can't get `puppet apply --catalog` to find files specified via `puppet://` when it runs, so I guess that plan is moot.

For posterity, here's where I got to:
```
export RUBYOPT=-W0 
FACTER_zulip_conf_path=/etc/zulip/zulip.conf FACTER_zulip_scripts_path=`pwd`/scripts ruby -rpuppet -e 'result = Puppet::Node::Facts.new(Puppet[:certname].to_s, Facter.to_hash)' -e 'result.add_local_facts' -e 'puts result.to_yaml' > /var/lib/puppet/yaml/facts/$(hostname --fqdn).yaml
puppet master --manifest <(echo 'Exec { path => "/usr/sbin:/usr/bin:/sbin:/bin" }'; echo "include zulip::profile::standalone") --modulepath=`pwd`/puppet:/srv/zulip-puppet-cache/current --compile `hostname --fqdn` > some-catalog.json

puppet apply --modulepath=`pwd`/puppet:/srv/zulip-puppet-cache/current --catalog new.catalog.json
```
```
Error: /Stage[main]/Zulip::Profile::Base/File[/etc/security/limits.conf]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///modules/zulip/limits.conf
Error: /Stage[main]/Zulip::Profile::Base/File[/usr/lib/nagios/plugins/zulip_base]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///modules/zulip/nagios_plugins/zulip_base
Notice: /Stage[main]/Zulip::Profile::Base/File[/usr/lib/nagios/plugins/zulip_base/check_debian_packages]: Dependency File[/usr/lib/nagios/plugins/zulip_base] has failures: true
Warning: /Stage[main]/Zulip::Profile::Base/File[/usr/lib/nagios/plugins/zulip_base/check_debian_packages]: Skipping because of failed dependencies
Error: /Stage[main]/Zulip::Nginx/File[/etc/nginx/zulip-include/]: Could not evaluate: Could not retrieve information from environment production source(s) puppet:///modules/zulip/nginx/zulip-include-common
Notice: /Stage[main]/Zulip::Nginx/File[/etc/nginx/zulip-include/api_headers]: Dependency File[/etc/nginx/zulip-include/] has failures: true
[...]

Reply:
OK, I've resolved this by making use of the [`tag` metaparameter](https://www.puppet.com/docs/puppet/5.5/metaparameter.html#tag) to install the hooks in a separate puppet run before anything else happens.  This adds an additional puppet run, but it's outside the critical period.  And it now means that hooks are matched in their pre- and post- action versions, and you get the code for both from the version you're deploying.